core/muxing: Redesign `StreamMuxer` trait

thomaseizinger opened the pull request:
# Description

<!-- Please write a summary of your changes and why you made them.-->

This PR re-designs the `StreamMuxer` trait to:

1. Use `&mut self` for all methods
2. Follow the style of "mutate state and return result via `poll`", same as we do for many other parts of `rust-libp2p`

## Splitting out separate PRs

- [x] https://github.com/libp2p/rust-libp2p/pull/2664
- [x] https://github.com/libp2p/rust-libp2p/pull/2685
- [x] https://github.com/libp2p/rust-libp2p/pull/2669
- [x] https://github.com/libp2p/rust-libp2p/pull/2665
- [x] https://github.com/libp2p/rust-libp2p/pull/2666
- [x] https://github.com/libp2p/rust-libp2p/pull/2667
- [x] https://github.com/libp2p/rust-libp2p/pull/2668
- [x] https://github.com/libp2p/rust-libp2p/pull/2677
- [x] https://github.com/libp2p/rust-yamux/pull/135
- [x] https://github.com/libp2p/rust-libp2p/pull/2688
- [x] https://github.com/libp2p/rust-libp2p/pull/2691
- [x] https://github.com/libp2p/rust-libp2p/pull/2703
- [x] https://github.com/libp2p/rust-libp2p/pull/2706
- [x] https://github.com/libp2p/rust-libp2p/pull/2707
- [x] https://github.com/libp2p/rust-libp2p/pull/2710

## Additional tasks

- [x] Retain `Substream` associated type on libp2p-core level
- [x] Fix mplex performance regression (use channels of `Bytes`/`[u8; X]`?)
- [ ] Make `StreamMuxer` use `Pin<&mut self>`
- [ ] ~Include "first close, then flush" logic within `SubstreamBox`~
- [ ] ~Release `yamux` once https://github.com/libp2p/rust-yamux/pull/135 is merged~
- [ ] ~Update PR with latest yamux release~
- [x] Drop b6c91d138b2d4078c2ba08b436256b4d035d59ab from PR now that cec117667b3e0a23b8acb3acbc342e087ef8df6b is landed and the bug is fixed.
- [ ] Write a more detailed changelog entry for `libp2p-yamux`.

## Links to any relevant issues

<!-- Reference any related issues.-->

## Open Questions

<!-- Unresolved questions, if any. -->

- ~Is there a better name for `AsyncReadWriteBox`?~

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Fix `StreamMuxer::Error` to io::Error

We are already enforcing that the associated type must convert to
`io::Error`. We might as well just make all functions return an
`io::Error` directly.

thomaseizinger commented on muxers/yamux/src/lib.rs:
```diff
@@ -84,116 +85,69 @@ where
+            .transpose()
+            .map_err(to_io_error)?
+            .map(|stream| StreamMuxerEvent::InboundSubstream(AsyncReadWriteBox::new(stream)))
+            .ok_or_else(|| to_io_error(yamux::ConnectionError::Closed));
```
Can you `?` here.

thomaseizinger commented on muxers/yamux/src/lib.rs:
```diff
@@ -84,116 +85,69 @@ where
+    }
+
+    fn poll(&mut self, c: &mut Context<'_>) -> Poll<io::Result<StreamMuxerEvent>> {
+        match self
```
This could be an `if let Poll::Ready`.

thomaseizinger commented on swarm/src/connection.rs:
```diff
@@ -79,7 +85,7 @@ where
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("Connection")
-            .field("muxing", &self.muxing)
+            // .field("muxing", &self.muxing)
```
Put this back in.

thomaseizinger reviewed (commented):
Things found in self-review.

mxinden commented on core/src/muxing.rs:
```diff
@@ -79,9 +79,6 @@ mod singleton;
 ///
 pub trait StreamMuxer {
-    /// Type of the object that represents the raw substream where data can be read and written.
-    type Substream;
```
What is the reasoning for using dynamic dispatching (`AsyncReadWriteBox`) instead of using an associated type that implements `AsyncRead` and `AsyncWrite`?

thomaseizinger replied:
Part of the motivation for this change was the idea that protocols (like `libp2p-ping`) should depend on a type that is independent from the specific muxer in question and instead are implemented on top of an opaque stream type. This would allow standalone usage of the protocol (state machines) which hopefully brings us closer to using some of the protocol logic independently from `libp2p-swarm`.

Additionally, I am not even sure an associated type would work unless we parametrize all protocols over `M: StreamMuxer` to be able to reference the concrete substream type in methods like `inject_fully_negotiated_inbound`?

thomaseizinger replied:
My current thinking is that it would be nice to have `libp2p-swarm` be an optional feature of protocol crates.

mxinden replied:
> My current thinking is that it would be nice to have `libp2p-swarm` be an optional feature of protocol crates.

Protocol crates like e.g. `libp2p-gossipsub`? Would they still be using `NetworkBehaviour`?

> Additionally, I am not even sure an associated type would work unless we parametrize all protocols over `M: StreamMuxer` to be able to reference the concrete substream type in methods like `inject_fully_negotiated_inbound`?

We might decide to do dynamic dispatching at the `libp2p-swarm` level, though I am not sure we need it at the `libp2p-core` level.

thomaseizinger replied:
> > My current thinking is that it would be nice to have `libp2p-swarm` be an optional feature of protocol crates.
> 
> Protocol crates like e.g. `libp2p-gossipsub`? Would they still be using `NetworkBehaviour`?

They would, optionally, by dispatching to standalone state machines that only depend on `libp2p-core` and could thus be used without the `libp2p-swarm` abstractions. Without derailing the conversation too much: I see one of two avenues for making `rust-libp2p` easier for "arbitrary stream usage".

1. Re-designing the `NetworkBehaviour` trait to provide more flexibility
2. Making it easier to not use `libp2p-swarm` but still be able to re-use the existing work and protocols.

> > Additionally, I am not even sure an associated type would work unless we parametrize all protocols over `M: StreamMuxer` to be able to reference the concrete substream type in methods like `inject_fully_negotiated_inbound`?
> 
> We might decide to do dynamic dispatching at the `libp2p-swarm` level, though I am not sure we need it at the `libp2p-core` level.

That is an interesting idea, I will go and explore that.

thomaseizinger replied:
What do yo think of https://github.com/libp2p/rust-libp2p/pull/2648/commits/61f07e0271e143992a00463f8d0ffe50b92bd32f?

mxinden replied:
:+1: for reintroducing `type Substream`.

mxinden commented on core/src/muxing.rs:
```diff
@@ -99,31 +104,13 @@ pub trait StreamMuxer {
-    ///
-    /// May panic or produce an undefined result if an earlier polling of the same substream
-    /// returned `Ready` or `Err`.
-    fn poll_outbound(
```
What if, as a user, I want to poll for outbound substreams, while at the same time signaling to the remote that I am not willing to accept any more inbound substreams? In other words, I do I exercise back-pressure on the number of inbound substreams?

I know this is not supported in mplex or yamux today. It is supported by QUIC and will be supported by https://github.com/libp2p/specs/pull/394.

thomaseizinger replied:
> while at the same time signaling to the remote that I am not willing to accept any more inbound substreams?

If this is a requirement, I'd suggest we make a function on `StreamMuxer` to trigger this behaviour.

In general though, I am a bit hesitant to do this. The original design of the `StreamMuxer` trait had numerous features that were completely unused (at least within the libp2p ecosystem that is known to me). Unless we use this functionality as part of `libp2p-swarm`, I'd refrain from adding it.

thomaseizinger replied:
> I know this is not supported in mplex or yamux today. It is supported by QUIC and will be supported by [libp2p/specs#394](https://github.com/libp2p/specs/pull/394).

This sounds like it would be better of as a method on the _concrete_ muxer type. Part of the benefits of the changes in this PR is that the `StreamMuxer` implementations are much easier to use standalone, meaning any usage outside of `libp2p-swarm` does not necessarily need to go through the `StreamMuxerBox` abstraction and hence can access functionality outside of the trait interface.

mxinden replied:
> > while at the same time signaling to the remote that I am not willing to accept any more inbound substreams?
> 
> If this is a requirement, I'd suggest we make a function on `StreamMuxer` to trigger this behaviour.

This will be a requirement. It isn't today, due to rust-libp2p not yet implementing a sophisticated stream multiplexer. Will follow up with more details.

thomaseizinger replied:
Fair point! Can we agree to design this once we have/add a sophisticated stream multiplexer? I am not confident that I can add this functionality now in a way that will be useful once we need it!

mxinden replied:
> I am not confident that I can add this functionality now in a way that will be useful once we need it!

Sounds good to me, especially since we don't prioritize outbound streams over inbound ones today.

https://github.com/libp2p/rust-libp2p/blob/3611d69e0911c5551270eaf8a6aa43d9b49f8782/swarm/src/connection/substream.rs#L154-L175

mxinden commented on swarm/src/connection/substream.rs:
```diff
@@ -18,196 +18,9 @@
-///
-/// The stream will close once both the inbound and outbound channels are closed, and no more
-/// outbound substream attempt is pending.
-pub struct Muxing<TMuxer, TUserData>
```
:rocket:

mxinden commented on muxers/yamux/src/lib.rs:
```diff
@@ -39,15 +39,7 @@ use std::{
 use thiserror::Error;
 
 /// A Yamux connection.
-pub struct Yamux<S>(Mutex<Inner<S>>);
```
Neat!

thomaseizinger replied:
Well, looking at the current build failures this may have been a mistake :D

mxinden commented on muxers/mplex/src/lib.rs:
```diff
@@ -236,3 +205,9 @@ impl AsyncWrite for Substream {
     }
 }
+
+impl Drop for Substream {
```
Is this change necessary for the general goal of _redesigning the `StreamMuxer` trait_? I would prefer unrelated changes to be split into separate pull requests, thus not having one very large one.

thomaseizinger replied:
I think it is coupled to it, yes. Where-else would you clean up the resources associated with a substream within the mplex implementation?

mxinden reviewed (commented):
Some very cool simplifications in here.

Any chance this pull request can be split into multiple autonomous ones to ease the review process?

thomaseizinger committed:
Add newline before return value

Visually separating this makes it easier to read.

thomaseizinger committed:
Re-design the `StreamMuxer` trait

The current `StreamMuxer` API is based around a poll-based approach
where each substream needs to be passed by value to read/write/close
it.

The muxer itself however is not available in the end-user abstractions
like `ConnectionHandler`. To circumvent this, the `StreamMuxerBox`
type exists which allows a `StreamMuxer` to be cloned. Together with
`SubstreamRef`, this allows each substream to own a reference to the
muxer it was created with and pass itself to the muxer within
implementations of `AsyncRead` and `AsyncWrite`. These implementations
are convenient as they allow an end-user to simply read and write to
the substream without holding a reference to anything else.

We can achieve the same goal by changing the `StreamMuxerBox` abstraction
to use a `SubstreamBox`. Users of `libp2p-core` can continue to
depend on the `Substream` associated type of `StreamMuxer`.

mxinden commented on muxers/mplex/tests/async_write.rs:
```diff
@@ -64,8 +64,8 @@ fn async_write() {
-        let mut buf = Vec::new();
-        outbound.read_to_end(&mut buf).await.unwrap();
+        let mut buf = vec![0u8; 11];
+        outbound.read_exact(&mut buf).await.unwrap();
```
Would that not defeat the purpose of the unit test?

Above:

>     // Tests that `AsyncWrite::close` implies flush.

Not using `read_to_end` thus no longer tests the closing and flushing, no?

thomaseizinger replied:
Yeah you might be right. I wasn't too happy about these changes but extracted them as a separate commit so we can discuss it easier.

I suspect there might be a bug somewhere but I couldn't spot it so far.

thomaseizinger replied:
I fixed it! The important commit is https://github.com/libp2p/rust-libp2p/commit/cec117667b3e0a23b8acb3acbc342e087ef8df6b.

This means I can drop this commit from the PR history but that will require a force-push. I've noted it in the PR description as a task for the final clean-up once this PR is closer to being merged :)

elenaf9 commented on core/src/muxing.rs:
```diff
@@ -51,178 +50,61 @@
     /// >           properly informing the remote, there is no difference between this and
     /// >           immediately dropping the muxer.
-    fn poll_close(&self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>>;
+    fn poll_close(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>>;
```
```suggestion
    fn poll_close(mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>>;
```
Why not move ownership if the muxer should not be used afterwards anyway?

thomaseizinger replied:
It is still a poll function that may return Pending and needs to be polled again!

elenaf9 replied:
Ahh yes of course you're right. Somehow that completely slipped my mind.

elenaf9 commented on core/src/muxing/async_read_write_box.rs:
```diff
@@ -0,0 +1,105 @@
+// Copyright 2022 Thomas Eizinger.
```
I think it would make sense to also move`StreamMuxerBox` and `Wrap` into this module, since they are only used in combination with each other. Maybe also rename the module to `boxed` analogous to `transport::boxed`, which does something similar for transports. 

> - Is there a better name for AsyncReadWriteBox?

`SubstreamBox`? Even though it is theoretically abstract and could be used independently of the Muxer logic, in practice we only use it to box the `Mutex::Substream`, so maybe just reflect that in the name?

thomaseizinger replied:
Excellent ideas! Will implement!

elenaf9 commented on muxers/yamux/CHANGELOG.md:
```diff
@@ -1,6 +1,9 @@
 # 0.37.0 [unreleased]
 
 - Update to `libp2p-core` `v0.33.0`.
+- Remove `YamuxResult` and `YamuxError` from the public API. See [PR 2648].
+
+[PR 2648]: https://github.com/libp2p/rust-libp2p/pull/2648
```
I think the changelog for yamux should also mention the muxer trait & substream handling changes. The current entry imo does not sufficiently inform users that this PR did internal changes that might affect performance or could have introduced bugs.

thomaseizinger replied:
This changelog entry will already be included within https://github.com/libp2p/rust-libp2p/pull/2664.

I'll write another a more detailed one, once https://github.com/libp2p/rust-libp2p/pull/2664 is merged and this is rebased on top of it!

elenaf9 reviewed (approved):
The muxer logic in libp2p is still new to me so I can't give a profound review right now, but from a high-level perspective I like the changes of this PR very much! The API is much more intuitive, and consistent with the rest of rust-libp2p.

mxinden commented on core/src/muxing.rs:
```diff
@@ -231,511 +114,92 @@ pub enum StreamMuxerEvent<T> {
-    ) -> Poll<Result<(), Self::Error>> {
-        self.inner.shutdown_substream(cx, s)
-    }
+    fn fetch_and_increment(&mut self) -> Self {
```
```suggestion
    fn get_then_increment(&mut self) -> Self {
```

Would be more intuitive to me. Not sure whether the current version is more idiomatic?

thomaseizinger replied:
Naming was inspired by `Atomic::fetch_add`: https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU8.html#method.fetch_add

Could probably just call it `fetch_add` directly?

mxinden replied:
Sounds good to me.

mxinden commented on core/src/muxing.rs:
```diff
@@ -231,511 +114,92 @@ pub enum StreamMuxerEvent<T> {
-        self.inner
-            .destroy_substream(list.remove(&substream).unwrap())
-    }
+        let stream = futures::ready!(poll_open())?;
```
This indirection seems counterintuitive to me. While more code, I would expect the muxer calling the `OutboundStreams` instead of the `OutboundStreams` calling the muxer to be easier to follow.

thomaseizinger replied:
The _concrete_ muxers are calling the `OutboundSubstreams` component.

mplex: https://github.com/thomaseizinger/rust-libp2p/blob/db7cb6b47ace2f844a7a6da6ece345d8d2f6657a/muxers/mplex/src/lib.rs#L118-L120
yamux: https://github.com/thomaseizinger/rust-libp2p/blob/db7cb6b47ace2f844a7a6da6ece345d8d2f6657a/muxers/yamux/src/lib.rs#L110-L113

thomaseizinger replied:
I am not sure how to otherwise design this abstraction so it can be reused by each muxer implementation.

mxinden replied:
I am more referring to the closer (`poll_open`) passed by the muxer to the `OutboundSubstreams`. Is that a strict requirement?

thomaseizinger replied:
I am not sure I completely follow but here is the thinking:

1. We should only poll to open a new open substream if the user requested one.
2. Requesting to open a new substream issues an `OutboundSubstreamId`.
3. Thus, opening a new substream is based on the state of having issued `OutboundSubstreamId`s
4. The correct IDs need to be mapped to the `OutboundSubstreamId`s that were handed out.

I could inline this abstraction but it felt a bit error prone to duplicate the behaviour of only polling for new substreams if we have issued IDs but not serviced them yet.

mxinden commented on muxers/yamux/src/lib.rs:
```diff
@@ -386,23 +333,16 @@ where
-            yamux::ConnectionError::Io(e) => e,
-            e => io::Error::new(io::ErrorKind::Other, e),
-        }
+fn to_io_error(e: yamux::ConnectionError) -> io::Error {
```
What do you think of upstreaming this to the `yamux` crate?

https://github.com/libp2p/rust-yamux

thomaseizinger replied:
We can upstream that yes!

thomaseizinger replied:
See https://github.com/libp2p/rust-yamux/pull/135.

mxinden commented on core/src/muxing.rs:
```diff
@@ -51,178 +50,62 @@
 ///
+/// A [`StreamMuxer`] is a stateful object and needs to be continuously polled to make progress.
 pub trait StreamMuxer {
-    /// Type of the object that represents the raw substream where data can be read and written.
```
Comment is roughly accurate, no?

thomaseizinger replied:
Correct. Removing this is a left-over from first removing the associated type and reintroducing it based on the review feedback :D

I'll restore the doc string as well!

mxinden reviewed (commented):
>  +583 −1,401 

Very cool! 

A couple of comments. Sorry, large change, thus I am still not all the way through.

mxinden commented on swarm/src/connection/substream.rs:
```diff
@@ -18,240 +18,9 @@
-use std::sync::Arc;
-use std::{fmt, io::Error as IoError, pin::Pin, task::Context, task::Poll};
-
 /// Endpoint for a received substream.
```
How about moving this into `connection.rs?

thomaseizinger replied:
Good idea!

mxinden commented on muxers/yamux/src/lib.rs:
```diff
@@ -84,110 +83,58 @@ where
-    ) -> Poll<YamuxResult<()>> {
-        Pin::new(s).poll_flush(c).map_err(|e| YamuxError(e.into()))
-    }
+        let event = ready!(self.incoming.poll_next_unpin(c))
```
Nit pick, instead of the `ready!`, how about following the style above using `if let Poll::Ready`. Say that we introduce a statement below this chain, we might miss the early return.

thomaseizinger replied:
Happy to do that!

thomaseizinger replied:
I did the same thing in mplex!

elenaf9 commented on core/src/muxing.rs:
```diff
@@ -51,178 +50,63 @@
-    /// The API of `OutboundSubstream` is totally opaque, and the object can only be interfaced
-    /// through the methods on the `StreamMuxer` trait.
-    fn open_outbound(&self) -> Self::OutboundSubstream;
+    fn poll(&mut self, cx: &mut Context<'_>)
+        -> Poll<io::Result<StreamMuxerEvent<Self::Substream>>>;
```
In what cases does it makes sense in poll functions to have `&mut self`, and when should it be `Pin<&mut Self>`? I wasn't sure which of the two to take in `Transport::poll`, and settled for `Pin<&mut Self>` to mirror the `Stream` trait. But it would be good to have more profound reasoning for when to use what in our API

thomaseizinger replied:
It is my understanding that `Pin` was introduced to allow self-referential structs in futures that are generated from async blocks. Within our hand-written state machines, we don't have self-referential structs and thus, they don't need to be pinned during execution within an executor. In other words, they will automatically implement `Unpin`.

Ideally, `Pin` wouldn't exist and all poll-functions would take `&mut self` but this wouldn't allow for safe self-referential structs within futures. Thus I think the way to go is: Use `&mut self` until the compiler complains :)

elenaf9 replied:
Thank you for the explanation @thomaseizinger!

> Thus I think the way to go is: Use &mut self until the compiler complains :)

But for traits we won't really know it the struct that implements it will have self-references, no? So the compiler will not complain when defining the trait, but might complain once a user implements it on their type.

thomaseizinger replied:
> > Thus I think the way to go is: Use &mut self until the compiler complains :)
> 
> But for traits we won't really know it the struct that implements it will have self-references, no? So the compiler will not complain when defining the trait, but might complain once a user implements it on their type. 

That is true!

The safer way would be to use `Pin` everywhere but it is also additional complexity. I could go either way.

There are probably no implementations of `StreamMuxer` outside of this repository so leaving it at `&mut self` shouldn't cause any problems but at the same time, the additional complexity would also only really affect us core maintainers 🤷‍♂️

mxinden replied:
In my eyes, for correctness reasons, we should mirror the `Future` standard of using `Pin<&mut Self>`. Also see https://github.com/libp2p/rust-libp2p/issues/1389.

thomaseizinger replied:
Interesting! Happy to move towards `Pin<&mut self>`. I guess without knowing, I was working towards #1389 :D

thomaseizinger commented on muxers/mplex/src/lib.rs:
```diff
@@ -46,30 +48,38 @@ impl UpgradeInfo for MplexConfig {
         future::ready(Ok(Multiplex {
-            io: Mutex::new(io::Multiplexed::new(socket, self)),
+            io: Arc::new(Mutex::new(io::Multiplexed::new(
+                SubstreamBox::new(socket),
```
@elenaf9 I remember why I didn't name it `SubstreamBox` initially! Because it is also used here to erase the type of the socket which would otherwise also show up in the type signature of `libp2p_mplex::Substream`.

However, I think that is actually not a problem because _now_ we box the substream later on anyway. In an earlier design we didn't!

mxinden commented on muxers/mplex/src/lib.rs:
```diff
@@ -79,117 +83,142 @@ where
+                substream,
+                outbound_id,
+            )));
+        }
```
We should favor local work over remote work, i.e. `self.io.lock()).poll_open_stream(cx)` should be polled before `self.io.lock().poll_next_stream(cx)`. Does that make sense in this scenario?

thomaseizinger replied:
Makes sense!

thomaseizinger replied:
Done in https://github.com/libp2p/rust-libp2p/pull/2648/commits/e9506aae7048184e90ceabd4dcbad3f07126308d.

thomaseizinger commented on muxers/mplex/benches/split_send_size.rs:
```diff
@@ -132,24 +130,16 @@ fn run(transport: &mut BenchTransport, payload: &Vec<u8>, listen_addr: &Multiadd
+            .unwrap()
+            .into_outbound_substream()
             .unwrap();
-        let mut off = 0;
-        loop {
-            let n = poll_fn(|cx| conn.write_substream(cx, &mut stream, &payload[off..]))
-                .await
-                .unwrap();
-            off += n;
-            if off == payload.len() {
-                poll_fn(|cx| conn.flush_substream(cx, &mut stream))
-                    .await
-                    .unwrap();
-                return;
-            }
-        }
```
@mxinden This is also differently implemented in the benchmark now but I can't see how this would impact performance much.

mxinden replied:
> but I can't see how this would impact performance much.

Me neither :/

thomaseizinger commented on muxers/mplex/src/lib.rs:
```diff
@@ -79,117 +83,142 @@ where
 /// and no [`StreamMuxerEvent::AddressChange`] event is ever emitted.
 pub struct Multiplex<C> {
-    io: Mutex<io::Multiplexed<C>>,
+    io: Arc<Mutex<io::Multiplexed<C>>>,
```
This is a `parkling_lot::Mutex`. I'll try if a native async one makes a difference.

thomaseizinger committed:
fixup! Re-design the `StreamMuxer` trait

Reduce diff.

thomaseizinger committed:
fixup! Re-design the `StreamMuxer` trait

Reduce diff.

thomaseizinger committed:
fixup! Re-design the `StreamMuxer` trait

Fix compile errors.

thomaseizinger committed:
fixup! Re-design the `StreamMuxer` trait

Fix docs error.
